server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -36,6 +40,21 @@
 
     public static final String BACKING_INDEX_PREFIX = ".ds-";
     public static final DateFormatter DATE_FORMATTER = DateFormatter.forPattern("uuuu.MM.dd");
+    // Datastreams' leaf readers should be sorted by desc order of their timestamp field, as it allows search time optimizations
+    public static Comparator<LeafReader> DATASTREAM_LEAF_READERS_SORTER =
+        Comparator.comparingLong(
+            (LeafReader r) -> {
+                try {
+                    PointValues points = r.getPointValues(DataStream.TimestampField.FIXED_TIMESTAMP_FIELD);
+                    if (points != null) {
+                        byte[] sortValue = points.getMaxPackedValue();
+                        return LongPoint.decodeDimension(sortValue, 0);
+                    }
+                } catch (IOException e) {
```

Comment:
added in a9d44d97d793246644ecc955b8889aef5cf86738

Reply:
@jimczi   I've opted to throw `IllegalStateException` instead of assertion.  This allows us to to return an arbitrary value as it was before.

I hope you are fine with the change.